cmd/kops-controller/pkg/nodebootstrap/server.go

```diff
@@ -0,0 +1,137 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	rest "k8s.io/client-go/rest"
+	"k8s.io/klog"
+	"k8s.io/kops/node-authorizer/pkg/server"
+	pb "k8s.io/kops/pkg/proto/nodebootstrap"
+)
+
+const (
+	// the namespace to place the secrets
+	tokenNamespace = "kube-system"
+)
+
+type nodeBootstrapService struct {
+	authorizer server.Authorizer
+
+	corev1Client corev1client.CoreV1Interface
+
+	options Options
+}
+
+// Options is the configuration for the NodeBootstrap server
+type Options struct {
+	TokenTTL time.Duration `json:"tokenTTL,omitempty"`
+}
+
+// PopulateDefaults sets the default configuration values
+func (o *Options) PopulateDefaults() {
+	o.TokenTTL = 5 * time.Minute
+}
+
+func NewNodeBootstrapService(restConfig *rest.Config, authorizer server.Authorizer, options *Options) (*nodeBootstrapService, error) {
+	s := &nodeBootstrapService{}
+
+	s.authorizer = authorizer
+
+	c, err := corev1client.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to kubernetes: %v", err)
+	}
+	s.corev1Client = c
+
+	s.options = *options
+
+	return s, nil
+}
+
+var _ pb.NodeBootstrapServiceServer = &nodeBootstrapService{}
+
+func (s *nodeBootstrapService) CreateKubeletBootstrapToken(ctx context.Context, request *pb.CreateKubeletBootstrapTokenRequest) (*pb.CreateKubeletBootstrapTokenResponse, error) {
+	peer, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("failed to get peer context")
+	}
+
+	nodeRegistration := &server.NodeRegistration{}
+	nodeRegistration.Spec.RemoteAddr = peer.Addr.String()
+
+	if err := s.authorizer.Authorize(ctx, nodeRegistration); err != nil {
+		// In general we prefer to log error details here, and only return minimal information to the client, because it may be an attacker
+		klog.Warningf("internal error during authorization: %v", err)
+		return nil, fmt.Errorf("internal error during authorization")
+	}
+
+	if !nodeRegistration.Status.Allowed {
+		// TODO: Use grpc Status errors?
+		return nil, fmt.Errorf("node registration not allowed")
+	}
+
+	token, err := s.createBootstrapToken(ctx, nodeRegistration)
+	if err != nil {
+		klog.Warningf("error creating bootstrap token: %v", err)
+		return nil, fmt.Errorf("error creating bootstrap token")
+	}
+
+	response := &pb.CreateKubeletBootstrapTokenResponse{}
+	if token != nil {
+		response.Token = &pb.Token{
+			BearerToken: token.ID + "." + token.Secret,
+		}
+	}
+
+	return response, nil
+}
+
+// createBootstrapToken generates a bootstrap token for the node, inserting it into k8s
+func (s *nodeBootstrapService) createBootstrapToken(ctx context.Context, request *server.NodeRegistration) (*server.Token, error) {
+	usages := []string{"authentication", "signing"}
+
+	// @step: generate a random token for them
+	token, err := server.NewToken()
+	if err != nil {
+		return nil, err
+	}
+
+	// @step: add the secret to the namespace
```

Comment:
Why is this storing the created token in a Secret? I'm concerned a compromised node could impersonate a node in a more privileged instancegroup. Or a pod with permission to read these could escalate to impersonating a node.

Reply:
Answering narrowly, this is how bootstrap tokens work.

Broadly, I do think your concerns are valid.  We can make our bootstrap tokens short lived, we can use the NodeAuthorizer to reduce impact, but it's still a concern.  The k8s secret also has all the information, whereas presumably it would be safer to store only the hash of the token (requiring an attacker to find a hash collision in a relatively short TTL).

Perhaps then we should move to just generating the kubelet TLS certificate, and trying to figure out what happens during rotation...  WDYT?

Reply:
Short lived doesn't help, as the compromised node could presumably set a watch for kube-system secrets and snatch it almost as soon as it's created.

But yes, this weakness appears to be designed in upstream of us. Just generating the TLS certificate might be simpler.